oauth2_http/java/com/google/auth/oauth2/IdToken.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2019, Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google LLC. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import java.io.Serializable;
+import java.util.Date;
+import java.util.Objects;
+
+import com.google.api.client.json.webtoken.JsonWebSignature;
+import com.google.common.base.MoreObjects;
+
+/**
+ * Represents a temporary IdToken and its JSONWebSingature object
+ */
+public class IdToken extends AccessToken implements Serializable {
+
+  private static final long serialVersionUID = -8514239465808977353L;
+
+  private final JsonWebSignature jws;
+
+  /**
+   * @param tokenValue     String representation of the Id token.
+   * @param jws            JsonWebSignature as object
+   */
+  public IdToken(String tokenValue, JsonWebSignature jws) {
```

Comment:
Everywhere we instantiate an `IdToken`, we are immediately preceding that line with `JsonWebSignature.parse(tokenValue)`. That logic should be handled in this constructor or via a static `IdToken.create(String)` method.

Reply:
yeah...thats what i initially wanted to do but ran into a bit of an issue:

since i'm extending `IdToken` as 
```
public class IdToken extends AccessToken
```

i need to `super()` in the constructor for `IdToken` which means i need to pass in

```java
public AccessToken(String tokenValue, Date expirationTime)
```

that is, i need to pass in the raw token value and a `Date` object defining the expiration time.

The only way i can get the expiration time out of the jwt is to parse the JWT first.

There maybe an easier way to work around this properly but this bit of awkwardness is due to  using AccessToken as a basis

Reply:
Can we do something like:
```java
public IdToken(String tokenValue) {
  try {
    this.jws = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, tokenValue);
  } catch (IOException e) {
    throw new IllegalArgumentException("Invalid JsonWebSignature token provided", e);
  }
  super(tokenValue, new Date(this.jws.getPayload().getExpirationTimeSeconds() * 1000));
}
```

or:
```java
private IdToken(String tokenValue, JsonWebSignature jws) {
  // same thing you have
}

public static IdToken create(String tokenValue) throws IOException {
  return create(tokenValue, OAuth2Utils.JSON_FACTORY);
}

public static IdToken create(String tokenValue, JsonFactory jsonFactory) throws IOException {
  return new IdToken(tokenValue, JsonWebSignature.parse(jsonFactory, tokenValue);
}
```

Reply:
ok, 

```java
  public IdToken(String tokenValue, JsonWebSignature jws) {
    super(tokenValue, new Date(jws.getPayload().getExpirationTimeSeconds() * 1000));
    this.jws = jws;
  }

  public static IdToken create(String tokenValue) throws IOException {
    return create(tokenValue, OAuth2Utils.JSON_FACTORY);
  }
  
  public static IdToken create(String tokenValue, JsonFactory jsonFactory) throws IOException {
    return new IdToken(tokenValue, JsonWebSignature.parse(jsonFactory, tokenValue));
  }  
```

works; i think in the ifrst one had the same issue with `super()` not getting called first